src/sasl/cyrsasl_external.erl

```diff
@@ -66,73 +76,150 @@ mech_step(#state{creds = Creds}, User) ->
         false ->
             {error, <<"not-authorized">>};
         true ->
-            do_mech_step(Creds, User)
+            NewCreds = maybe_add_auth_id(Creds, User),
+            do_mech_step(NewCreds)
     end.
-do_mech_step(Creds, User) ->
-    XmppAddrs = get_credentials(Creds, xmpp_addresses),
-    CommonName = get_credentials(Creds, common_name),
+
+%%%=============================================================================
+%%%  local functions
+%%%=============================================================================
+start_all_modules() ->
+    Modules = [M || H <- ?MYHOSTS, {mod, M} <- get_verification_list(H)],
+    case code:ensure_modules_loaded(Modules) of
+        {error, Error} -> error(Error);
+        _ -> ok
+    end,
+    lists:map(fun start_module/1, Modules).
+
+stop_all_modules() ->
+    [stop_module(M) || H <- ?MYHOSTS, {mod, M} <- get_verification_list(H)].
+
+start_module(Module) -> run_module_fn(Module, start).
+
+stop_module(Module) -> run_module_fn(Module, stop).
+
+run_module_fn(Module, Fn) ->
+    case erlang:function_exported(Module, Fn, 0) of
+        true -> Module:Fn();
+        _ -> ok
+    end.
+
+get_common_name(Cert) ->
+    case cert_utils:get_common_name(Cert) of
+        error -> [];
+        CN -> [{common_name, CN}]
+    end.
+
+get_xmpp_addresses(Cert) ->
```

Comment:
I might be wrong, but:
```erlang
get_xmpp_addresses(Cert) ->
    XmmpAddresses = cert_utils:get_xmpp_addresses(Cert),
    [{xmpp_addresses, XmmpAddresses}].
```
should do the same, since whatever is the result of case it is put as a second argument in a tuple in list.